src/attr.c

```diff
@@ -322,12 +322,12 @@ static int attr_setup(git_repository *repo, git_attr_session *attr_session)
 	 * definitions will be available for later file parsing.
 	 */
 
-	if ((error = system_attr_file(&path, attr_session)) < 0 ||
-	    (error = preload_attr_file(repo, attr_session, GIT_ATTR_FILE__FROM_FILE,
-				       NULL, path.ptr)) < 0) {
-		if (error != GIT_ENOTFOUND)
-			goto out;
-	}
+	error = system_attr_file(&path, attr_session);
+	if (error == 0)
+		error = preload_attr_file(repo, attr_session, GIT_ATTR_FILE__FROM_FILE,
+					  NULL, path.ptr);
+	if (error != GIT_ENOTFOUND)
+		goto out;
```

Comment:
Yup, this indeed did change behaviour. Previous to my commit 5452e49fce21f726bec19519da7f012e3f19e736, we would've returned if ``preload_attr_file` returned any value except `GIT_ENOTFOUND`. Now we only return if there was an error, but will not return if we were successful. 

But was previous behaviour really the correct thing to do? The intention of `attr_setup` is to pre-cache all attribute files that are relevant to the current repository. So we want to load the global attributes file, the repo attributes file as well as attributes files from the working directory and index. Previously, we would've returned as soon as we have successfully preloaded the system attributes file without preloading any of the others.

You say that the fix here is that `preload_attr_file` should only be called if `error == 0`. But this is indeed not true: `system_attr_file` may only return either a negative value or 0, there is no codepath that returns positive. So with my refactorings, `preload_attr_file` is _still_ called if and only if `error == 0`. So the real change you do here is the return on success.

So I'm not yet convinced that the old behaviour is the correct thing to do. Could you please provide a reproducer with the changing behaviour you have encountered? Just want to make up my own mind whether your proposed fix is the correct thing to do or if you were previously relying on libgit2's faulty behaviour.

Reply:
I already posted a case where this failed.

Reply:
Considering to change a behavior is fine, but it should not be called "Refactoring", but it should also not break things.

Reply:
Btw in TortoiseGit I also have tests which rely on a working .gitattributes files (e.g., with a filter) and this is working correctly with libgit2 even with the early exit here.

Reply:
In my case w/o the early exit `attr_setup` fails here: https://github.com/libgit2/libgit2/blob/master/src/attr.c#L340 and returns -1

Reply:
I just found the reason: `git_buf_sets` fails in `git_repository_item_path` if the buffer is not empty. Comming up with a new PR if my tests run w/o errors now.

Reply:
@pks-t Commit pushed.

Reply:
Btw. I just checked the coude again and the early exit was fine, because the "other" files are loaded *again* after the call to `attr_setup` on line https://github.com/libgit2/libgit2/blob/master/src/attr.c#L515.

Reply:
> I already posted a case where this failed: #4967 (comment)

Oops, sorry, didn't see that one.

> Considering to change a behavior is fine, but it should not be called "Refactoring" and it should also not break things.

I fully agree with you. The change in behaviour definitely wasn't intended, even though it _seems_ to me like the right thing to do in hindsight. But as said, I don't yet know if I'm correct about that.

> I just found the reason: git_buf_sets fails in git_repository_item_path if the buffer is not empty. Comming up with a new PR if my tests run w/o errors now.

Ah, interesting.

> Btw. I just checked the coude again and the early exit was fine, because the "other" files are loaded again after the call to attr_setup on line https://github.com/libgit2/libgit2/blob/master/src/attr.c#L515.

That's... weird. The current code could be improved by quite a lot, if you ask me. Either we do the loading in `attr_setup`, or if we load the files anyway after `attr_setup` then we should just inline the one file we were _not_ loading in `attr_setup`, which is the global one. After all, `attr_setup` is used in this one location, only.

I'll take a look at your changes and try to dig into that code a bit more.